apps/site/src/components/pages/hub/block-data-container/sandboxed-block-demo.tsx

```diff
@@ -0,0 +1,216 @@
+import { ServiceEmbedderMessageCallbacks } from "@blockprotocol/service";
+import { ExternalServiceMethodRequest } from "@local/internal-api-client/api";
+import {
+  FunctionComponent,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+
+import { apiClient } from "../../../../lib/api-client";
+import { ExpandedBlockMetadata } from "../../../../lib/blocks";
+import { ServiceModuleModal } from "./sandboxed-block-demo/service-module-modal";
+
+export interface SandboxedBlockProps {
+  metadata: ExpandedBlockMetadata;
+  props: Record<string, unknown> | undefined;
+  sandboxBaseUrl: string;
+}
+
+const generateBlockServicePermissionKey = (blockPath: string) =>
+  `allow-${blockPath}-service-usage`;
+
+const setServiceUsagePermission = (blockPath: string, allowed: boolean) => {
+  sessionStorage.setItem(
+    generateBlockServicePermissionKey(blockPath),
+    allowed.toString(),
+  );
+};
```

Comment:
This doesn't block merging, but I wonder if we just want to save this in the page's state instead. If I'm browsing through the blocks the hub, make the decision to not allow requests from a block, continue browsing the hub in the same tab and then later change my mind, it's not obvious how I go about changing my preference.

At that point it's also not obvious that requests are failing, and that that's why the block isn't functioning as intended.

Perhaps we could either store the preference in the pages state so that it isn't persisted across redirects in the same tab, or we can have some kind of message fade-in under the block reminding the user that the external service message requests are being blocked because of a decision they made previously, ideally with something like an "Undo" button?

Reply:
It's a fair point – I've created a [follow-up task](https://app.asana.com/0/1203623179643290/1204138539312008/f) to do this tomorrow